### PR TITLE
Fourth-module/third-task

### PR DESCRIPTION
**Pull request by PrimoTibalt**

### CODE REVIEW

File: ModuleFourTasks/Task3/UserTaskService.cs [hunk 1]
```diff
@@ -12,25 +12,29 @@ public UserTaskService(IUserDao userDao)
             _userDao = userDao;
         }
 
-        public int AddTaskForUser(int userId, UserTask task)
+        public void AddTaskForUser(int userId, UserTask task)
         {
             if (userId < 0)
-                return -1;
+            {
+                throw new ArgumentException("Invalid userId");
```
Comment (alexander-trishin) [inline ModuleFourTasks/Task3/UserTaskService.cs]: More appropriate exception here will be `ArgumentOutOfRangeException`

File: ModuleFourTasks/Task3/UserTaskService.cs [hunk 2]
```diff
@@ -12,25 +12,29 @@ public UserTaskService(IUserDao userDao)
...
+            }
 
             var user = _userDao.GetUser(userId);
             if (user == null)
-                return -2;
+            {
+                throw new ArgumentException("User not found");
```
Comment (alexander-trishin) [inline ModuleFourTasks/Task3/UserTaskService.cs]: Please, create a custom exception `UserNotFoundException`. [How to create user-defined exceptions](https://docs.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions)

File: ModuleFourTasks/Task3/UserTaskService.cs [hunk 3]
```diff
@@ -12,25 +12,29 @@ public UserTaskService(IUserDao userDao)
...
+            }
 
             var tasks = user.Tasks;
             foreach (var t in tasks)
             {
                 if (string.Equals(task.Description, t.Description, StringComparison.OrdinalIgnoreCase))
-                    return -3;
+                {
+                    throw new ArgumentException("The task already exists");
```
Comment (alexander-trishin) [inline ModuleFourTasks/Task3/UserTaskService.cs]: You can also use `InvalidOperationException`

**Review by alexander-trishin — CHANGES_REQUESTED** [review]
Hi Anton,

A few advises about exception handling:

1. Choose a type (Exception-derived type) that is most meaningful. This will give a chance for the code up in the call stack to appropriately handle the specific case. In case there is no .Net exception that matches exact semantics, create a custom one.

2. Incase you create a hierarchy of Custom exceptions. It's best to keep the hierarchy shallow. So as to create few base classes, because base classes act as a way of treating all errors as one error. On the same lines, never throw `System.Exception`

3. Do keep the message of the exception detailed enough for the developer to fix, in case the exception goes unhandled. At the same time, too much information to outside world can be a security issue.

File: ModuleFourTasks/Task3/UserTaskService.cs (hunk 3 above)
Comment (PrimoTibalt) [reply ModuleFourTasks/Task3/UserTaskService.cs]: Fixed in cbd4ea3.

File: ModuleFourTasks/Task3/UserTaskService.cs (hunk 2 above)
Comment (PrimoTibalt) [reply ModuleFourTasks/Task3/UserTaskService.cs]: Fixed in cbd4ea3.

File: ModuleFourTasks/Task3/UserTaskService.cs (hunk 1 above)
Comment (PrimoTibalt) [reply ModuleFourTasks/Task3/UserTaskService.cs]: Fixed in cbd4ea3.

**Review by alexander-trishin — APPROVED** [review]

**Pull request merged**